File: implementations/python/src/aep/mcp_bridge.py
import json
import threading

from .harness import AepHarness
from .task import TaskTracker
from .errors import ErrorCode


class McpBridge:
    def __init__(self, aep=None, transport=None, server_info=None):
        self.aep = aep or AepHarness()
        self.transport = transport
        self.tools = {}
        self.server_info = server_info or {"name": "aep-mcp-bridge", "version": "0.1.0"}

    def register_tool(self, tool_def: dict):
        self.tools[tool_def["name"]] = {
            "handler": tool_def["handler"],
            "schema": tool_def.get("schema", {}),
        }
        return self

    def handle_request(self, request: dict):
        if not request or not request.get("method"):
            return self._error(None, -32600, "Invalid Request")
        method = request["method"]
        if method == "initialize":
            return self._handle_initialize(request)
        if method == "notifications/initialized":
            return None
        if method == "tools/list":
            return self._handle_tools_list(request)
        if method == "tools/call":
            return self._handle_tools_call(request)
        return self._error(request.get("id"), -32601, f"Method not found: {method}")

    def _handle_initialize(self, request):
        return {
            "jsonrpc": "2.0",
            "id": request.get("id"),
            "result": {
                "protocolVersion": "0.1.0",
                "capabilities": {"tools": {}},
                "serverInfo": self.server_info,
            },
        }

    def _handle_tools_list(self, request):
        tool_list = []
        for name, entry in self.tools.items():
            schema = entry["schema"]
            tool_list.append({
                "name": name,
                "description": schema.get("description", f"AEP-backed tool: {name}"),
                "inputSchema": {
                    "type": "object",
                    "properties": schema.get("properties", {}),
                    "required": schema.get("required", []),
                },
            })
        return {"jsonrpc": "2.0", "id": request.get("id"), "result": {"tools": tool_list}}

    def _handle_tools_call(self, request):
        params = request.get("params") or {}
        name = params.get("name")
        args = params.get("arguments") or {}
        if name not in self.tools:
            return self._error(request.get("id"), -32602, f"Unknown tool: {name}")
        handler = self.tools[name]["handler"]
        try:
            result = handler(args, {
                "aep": self.aep,
                "transport": self.transport,
                "task_id": args.get("_task_id"),
                "session_id": args.get("_session_id"),
            })
            return {"jsonrpc": "2.0", "id": request.get("id"), "result": result}
        except Exception as err:
            return {"jsonrpc": "2.0", "id": request.get("id"), "result": {
                "isError": True,
                "content": [{"type": "text", "text": str(err)}],
            }}

    def _error(self, request_id, code, message):
        return {"jsonrpc": "2.0", "id": request_id, "error": {"code": code, "message": message}}


def async_tool_handler(name: str, options: dict):
    description = options.get("description")
    input_schema = options.get("input_schema", {})
    work = options["work"]

    def handler(args, ctx):
        tracker = TaskTracker(
            task_id=args.get("_task_id") or None,
            source=f"tool:{name}",
            session_id=ctx.get("session_id"),
            description=json.dumps(args),
        )
        transport = ctx.get("transport")

        def _send(event):
            if transport is not None:
                transport.send(event)

        _send(tracker.accepted())

        def _run():
            try:
                _send(tracker.started())
                _send(tracker.progress({"progress": 0.5, "message": f"{name} in progress"}))
                result = work(args, {"tracker": tracker, "ctx": ctx})
                _send(tracker.completed(result))
            except Exception as err:
                _send(tracker.failed(ErrorCode.TOOL_ERROR, str(err)))

        threading.Thread(target=_run, daemon=True).start()

        return {"content": [{"type": "text", "text": json.dumps({"task_id": tracker.id, "status": "accepted"})}]}

    return {
        "name": name,
        "schema": {
            "description": description,
            "properties": input_schema.get("properties", {}),
            "required": input_schema.get("required", []),
        },
        "handler": handler,
    }

File: implementations/typescript/src/bridge/mcp-bridge.js
import { HarmovelaHarness, TaskTracker, ErrorCode, errorPayload } from "../index.js";

export class McpBridge {
  constructor(options = {}) {
    this.aep = options.aep ?? new HarmovelaHarness();
    this.transport = options.transport ?? null;
    this.tools = new Map();
    this.serverInfo = options.serverInfo ?? { name: "aep-mcp-bridge", version: "0.1.0" };
    this._initialized = false;
    this._clientCapabilities = null;
  }

  registerTool(toolDef) {
    this.tools.set(toolDef.name, { handler: toolDef.handler, schema: toolDef.schema });
    return this;
  }

  async handleRequest(request) {
    if (!request || !request.method) {
      return this._errorResponse(null, -32600, "Invalid Request");
    }

    switch (request.method) {
      case "initialize": return this._handleInitialize(request);
      case "notifications/initialized": return null;
      case "tools/list": return this._handleToolsList(request);
      case "tools/call": return this._handleToolsCall(request);
      default: return this._errorResponse(request.id, -32601, `Method not found: ${request.method}`);
    }
  }

  _handleInitialize(request) {
    this._initialized = true;
    this._clientCapabilities = request.params?.capabilities ?? {};

    return {
      jsonrpc: "2.0",
      id: request.id,
      result: {
        protocolVersion: "0.1.0",
        capabilities: {
          tools: {}
        },
        serverInfo: this.serverInfo
      }
    };
  }

  _handleToolsList(request) {
    const toolList = [];
    for (const [name, { schema }] of this.tools) {
      toolList.push({
        name,
        description: schema.description ?? `AEP-backed tool: ${name}`,
        inputSchema: {
          type: "object",
          properties: schema.properties ?? {},
          required: schema.required ?? []
        }
      });
    }

    return {
      jsonrpc: "2.0",
      id: request.id,
      result: { tools: toolList }
    };
  }

  async _handleToolsCall(request) {
    const { name, arguments: args } = request.params ?? {};

    if (!this.tools.has(name)) {
      return this._errorResponse(request.id, -32602, `Unknown tool: ${name}`);
    }

    const { handler } = this.tools.get(name);

    try {
      const result = await handler(args, {
        aep: this.aep,
        transport: this.transport,
        taskId: args?._task_id,
        sessionId: args?._session_id
      });
      return { jsonrpc: "2.0", id: request.id, result };
    } catch (err) {
      return { jsonrpc: "2.0", id: request.id, result: {
        isError: true,
        content: [{ type: "text", text: err.message }]
      }};
    }
  }

  _errorResponse(id, code, message) {
    return {
      jsonrpc: "2.0",
      id: id ?? null,
      error: { code, message }
    };
  }
}

export function asyncToolHandler(name, { description, inputSchema = {}, work }) {
  return {
    name,
    schema: {
      description,
      properties: inputSchema.properties ?? {},
      required: inputSchema.required ?? []
    },
    handler: async (args, ctx) => {
      const tracker = new TaskTracker({
        task_id: args._task_id ?? `task_${Date.now().toString(36)}`,
        source: `tool:${name}`,
        session_id: ctx.sessionId,
        description: JSON.stringify(args)
      });

      const accepted = tracker.accepted();
      ctx.transport?.send(accepted);

      // Execute in background
      setImmediate(async () => {
        try {
          const started = tracker.started();
          ctx.transport?.send(started);

          const progress = tracker.progress({ progress: 0.5, message: `${name} in progress` });
          ctx.transport?.send(progress);

          const result = await work(args, { tracker, ctx });

          const completed = tracker.completed(result);
          ctx.transport?.send(completed);
        } catch (err) {
          const failed = tracker.failed(ErrorCode.TOOL_ERROR, err.message);
          ctx.transport?.send(failed);
        }
      });

      return {
        content: [{ type: "text", text: JSON.stringify({ task_id: tracker.id, status: "accepted" }) }]
      };
    }
  };
}
